smack-extensions/src/main/java/org/jivesoftware/smackx/usertune/element/UserTuneElement.java

```diff
@@ -0,0 +1,152 @@
+    public static final String ELEMENT = "tune";
+
+    private final String artist;
+    private final int length;
+    private final int rating;
+    private final String source;
+    private final String title;
+    private final int track;
+    private final URI uri;
+
+    private UserTuneElement(UserTuneElementBuilder builder) {
+        this.artist = builder.artist;
+        this.length = builder.length;
+        this.rating = builder.rating;
+        this.source = builder.source;
+        this.title = builder.title;
+        this.track = builder.track;
+        this.uri = builder.uri;
+    }
+
+    @Override
+    public String getNamespace() {
+        return NAMESPACE;
+    }
+
+    @Override
+    public String getElementName() {
+        return ELEMENT;
+    }
+
+    @Override
+    public XmlStringBuilder toXML(XmlEnvironment xmlEnvironment) {
+        XmlStringBuilder xml = new XmlStringBuilder(this);
+        if (isNullUserTune()) {
+            return xml.closeEmptyElement();
+        }
+        xml.rightAngleBracket();
+        xml.optAppend(getUserTuneXML());
+        return xml.closeElement(getElementName());
+    }
+
+    private CharSequence getUserTuneXML() {
+        String userTuneXML = "<artist>" + artist + "</artist>" +
+                "<length>" + length + "</length>" +
+                "<rating>" + rating + "</rating>" +
+                "<source>" + source + "</source>" +
+                "<title>" + title + "</title>" +
+                "<track>" + track + "</track>" +
+                "<uri>" + uri + "</uri>";
+        return userTuneXML;
+    }
+
+    private boolean isNullUserTune() {
+        if (this.artist == null && this.length == 0 && this.rating == 0 && this.source == null &&
+            this.title == null && this.track == 0 && this.uri == null) {
+            return true;
+        }
+        else {
+            return false;
+        }
+    }
+
+    public static boolean hasUserTuneElement(Message message) {
+        return message.hasExtension(ELEMENT,NAMESPACE);
+    }
+
+    public static UserTuneElement fromMessage(Message message) {
+        return message.getExtension(ELEMENT,NAMESPACE);
+    }
+
+    public UserTuneElement getUserTune() {
+        return this;
+    }
+
+    public static class UserTuneElementBuilder{
+        private String artist;
+        private int length;
+        private int rating;
+        private String source;
+        private String title;
+        private int track;
+        private URI uri;
+
+        public void setArtist(String artist) {
+            this.artist = artist;
+        }
+
+        public void setLength(int length) {
+            this.length = length;
+        }
+
+        public void setRating(int rating) {
+            if (rating <= 10 && rating >= 0) {
+                this.rating = rating;
+            }
+            else {
+                throw new IllegalArgumentException("Rating tried to set was out of bounds");
```

Comment:
This would be a good opportunity to inform the user, which bounds they should stay inside ;)

Reply:
Proper documentation added to inform user about the rating range.